contracts/registries/LicenseRegistry.sol

```diff
@@ -1,64 +1,78 @@
-    mapping(bytes32 licenseHash => uint256 licenseId) private _hashedLicenses;
-
-    /// @dev Maps the licenseId to the license data
-    mapping(uint256 licenseId => Licensing.License licenseData) private _licenses;
+    /// @dev Storage of the LicenseRegistry
+    /// @param name Name of the Programmable IP License NFT
+    /// @param symbol Symbol of the Programmable IP License NFT
+    /// @param imageUrl URL of the Licensing Image
+    /// @param licensingModule Returns the canonical protocol-wide LicensingModule
+    /// @param disputeModule Returns the canonical protocol-wide DisputeModule
+    /// @param hashedLicenses Maps the hash of the license data to the licenseId
+    /// @param licenses Maps the licenseId to the license data
+    /// @param mintedLicenses Tracks the number of licenses registered in the protocol,
+    /// it will not decrease when a license is burnt.
+    /// @custom:storage-location erc7201:story-protocol.LicenseRegistry
+    struct LicenseRegistryStorage {
+        string name;
+        string symbol;
+        string imageUrl;
+        ILicensingModule licensingModule;
+        IDisputeModule disputeModule;
+        mapping(bytes32 licenseHash => uint256 licenseId) hashedLicenses;
+        mapping(uint256 licenseId => Licensing.License licenseData) licenses;
+        uint256 mintedLicenses;
+    }
 
-    /// @dev Tracks the number of licenses registered in the protocol, it will not decrease when a license is burnt.
-    uint256 private _mintedLicenses;
+    // keccak256(abi.encode(uint256(keccak256("story-protocol.LicenseRegistry")) - 1)) & ~bytes32(uint256(0xff));
+    bytes32 private constant LicenseRegistryStorageLocation =
+        0x5ed898e10dedf257f39672a55146f3fecade9da16f4ff022557924a10d60a900;
 
     /// @dev We have to implement this modifier instead of inheriting `LicensingModuleAware` because LicensingModule
     /// constructor requires the licenseRegistry address, which would create a circular dependency. Thus, we use the
     /// function `setLicensingModule` to set the licensing module address after deploying the module.
     modifier onlyLicensingModule() {
-        if (msg.sender != address(LICENSING_MODULE)) {
+        if (msg.sender != address(_getLicenseRegistryStorage().licensingModule)) {
             revert Errors.LicenseRegistry__CallerNotLicensingModule();
         }
         _;
     }
 
-    constructor(address governance, string memory url) ERC1155("") Governable(governance) {
-        imageUrl = url;
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(address governance, string memory imageUrl) public initializer {
+        __ERC1155_init("");
+        __GovernableUpgradeable_init(governance);
+        __UUPSUpgradeable_init();
+        LicenseRegistryStorage storage $ = _getLicenseRegistryStorage();
+        $.name = "Programmable IP License NFT";
```

Comment:
is this a new coding style to use "$"?

Reply:
Yes, $ is the nomenclature suggested in [ERC7201](https://eips.ethereum.org/EIPS/eip-7201) to identify the struct that holds the storage for the contract, which is place in the namespaced storage slot:

```pragma solidity ^0.8.20;

contract Example {
    /// @custom:storage-location erc7201:example.main
    struct MainStorage {
        uint256 x;
        uint256 y;
    }

    // keccak256(abi.encode(uint256(keccak256("example.main")) - 1)) & ~bytes32(uint256(0xff));
    bytes32 private constant MAIN_STORAGE_LOCATION =
        0x183a6125c38840424c4a85fa12bab2ab606c4b6d0e7cc73c0c06ba5300eab500;

    function _getMainStorage() private pure returns (MainStorage storage $) {
        assembly {
            $.slot := MAIN_STORAGE_LOCATION
        }
    }

    function _getXTimesY() internal view returns (uint256) {
        MainStorage storage $ = _getMainStorage();
        return $.x * $.y;
    }
}
```